[RFC] Adds countdown loops RFC

damirka opened the pull request:
RFC for #554.

gluax commented on docs/rfc/005-countdown-loops.md:
```diff
@@ -0,0 +1,98 @@
+# Motivation
+
+In the current design of the language only incremental ranges are allowed. Though
+in some cases there's a need for loops going in reverse direction. This example
```
going in the reverse direction

gluax commented on docs/rfc/005-countdown-loops.md:
```diff
@@ -0,0 +1,98 @@
+
+In the current design of the language only incremental ranges are allowed. Though
+in some cases there's a need for loops going in reverse direction. This example
+demonstrates bubble sort algorithm where countdown loops are mocked:
```
demonstrates the

gluax commented on docs/rfc/005-countdown-loops.md:
````diff
@@ -0,0 +1,98 @@
+}
+```
+
+Having countdown loop in the example above could improve readability and 
````
Having a countdown loop

gluax commented on docs/rfc/005-countdown-loops.md:
```diff
@@ -0,0 +1,98 @@
+Having countdown loop in the example above could improve readability and 
+usability of the language by making it more natural to the developer.
+
+However, if we imagined this example using countdown loop, we would see that 
```
using a countdown loop

gluax commented on docs/rfc/005-countdown-loops.md:
```diff
@@ -0,0 +1,98 @@
+usability of the language by making it more natural to the developer.
+
+However, if we imagined this example using countdown loop, we would see that 
+it wouldn't be possible to count to 0; because first bound of the range is
```
because the first bound

gluax commented on docs/rfc/005-countdown-loops.md:
```diff
@@ -0,0 +1,98 @@
+
+However, if we imagined this example using countdown loop, we would see that 
+it wouldn't be possible to count to 0; because first bound of the range is
+inclusive and second is exclusive, and loops ranges must use only unsigned integers.
```
and the second

gluax commented on docs/rfc/005-countdown-loops.md:
```diff
@@ -0,0 +1,98 @@
+## Inclusive ranges
+
+To solve loop asymmetry and to improve loop ranges in general we suggest adding 
+inclusive range operator to Leo. Inclusive range would extend second bound of the
```
adding an inclusive range, extend the second bound

gluax reviewed (approved):
Grammar fixes, but otherwise LGTM!

acoglio reviewed (approved):
Looks good.

acoglio commented:
@gluax's suggestions look good too.

bendyarm reviewed (approved):
Looks good, and gluax' edits look good, too.
Suggestion: in the # Design section, add a 3rd subsection called:
## Example
where the bubble sort is rewritten using the new feature.

acoglio commented:
I like @bendyarm's suggestion of the bubble sort example.

damirka commented:
@acoglio @bendyarm @gluax I have taken a closer look at the bubble sort example that we had and there was no way (correct me if I'm wrong) to rewrite it with countdown loops. Instead I added fixed round (we don't have break or continue statements) shaker sort algorithm which demonstrates the problem and the solution.

And added example to the Design section.

damirka commented:
Oh. And by the way it does solve the problem in smaller number of constraints (~20% less). 😆

acoglio commented:
@damirka The edits and the shaker sort example look good.

For bubble sort, the idea was to flip the outer loop in https://github.com/AleoHQ/leo/issues/554:
```
function bubble_sort(mut a: [u32; 10]) -> [u32; 10] {
    for i in 9..0 { // counts down
        for j in 0..i {
            if (a[j] > a[j+1]) {
                let tmp = a[j];
                a[j] = a[j+1];
                a[j+1] = tmp;
            }
        }
    }
    return a
}
```
That is, `i` is the position where the largest element of the subarray `a[0..i-1]` "bubbles up" into, at every outer loop's iteration. Does that look right? If it does, we could include the bubble sort example along with the shaker sort example in the RFC, or we perhaps shaker sort is enough. Bubble sort is slightly simpler than shaker sort though. Up to you. @bendyarm, any thoughts?

acoglio commented:
I'm merging this now. We can add bubble sort later if we want.
